pennylane/templates/subroutines/semi_adder.py

```diff
@@ -0,0 +1,289 @@
+            y_wires (Sequence[int]): the wires that store the integer :math:`y`
+            work_wires (Sequence[int]): the auxiliary wires to use for the addition. ``len(y_wires) . 1`` work
+                wires should be provided.
+
+        Returns:
+            list[.Operator]: Decomposition of the operator
+        """
+        op_list = []
+
+        # revert wires to follow PennyLane convention
+
+        x_wires_pl = x_wires[::-1][: len(y_wires)]
+        y_wires_pl = y_wires[::-1]
+        work_wires_pl = work_wires[::-1]
+        op_list.append(qml.TemporaryAnd([x_wires_pl[0], y_wires_pl[0], work_wires_pl[0]]))
+
+        for i in range(1, len(y_wires_pl) - 1):
+            if i < len(x_wires_pl):
+                op_list += _left_operator(
+                    [work_wires_pl[i - 1], x_wires_pl[i], y_wires_pl[i], work_wires_pl[i]]
+                )
+            else:
+                # If the number of qubits in |x> is smaller than |y>, we can conceptually complete the |x> state
+                # with |0> on the left making sure they are of the same size. Assuming this, we can simplify the left operator.
+                op_list.append(qml.CNOT([work_wires_pl[i - 1], y_wires_pl[i]]))
+                op_list.append(
+                    qml.TemporaryAnd([work_wires_pl[i - 1], y_wires_pl[i], work_wires_pl[i]])
+                )
+                op_list.append(qml.CNOT([work_wires_pl[i - 1], work_wires_pl[i]]))
+
+        op_list.append(qml.CNOT([work_wires_pl[-1], y_wires_pl[-1]]))
+
+        if len(x_wires_pl) >= len(y_wires_pl):
+            op_list.append(qml.CNOT([x_wires_pl[-1], y_wires_pl[-1]]))
+
+        for i in range(len(y_wires_pl) - 2, 0, -1):
+            if i < len(x_wires_pl):
+                op_list += _right_operator(
+                    [work_wires_pl[i - 1], x_wires_pl[i], y_wires_pl[i], work_wires_pl[i]]
+                )
+            else:
+                # If the number of qubits in |x> is smaller than |y>, we can conceptually complete the |x> state
+                # with |0> on the left making sure they are of the same size. Assuming this, we can simplify the right operator.
+                op_list.append(qml.CNOT([work_wires_pl[i - 1], work_wires_pl[i]]))
+                op_list.append(
+                    qml.adjoint(
+                        qml.TemporaryAnd([work_wires_pl[i - 1], y_wires_pl[i], work_wires_pl[i]])
+                    )
+                )
+
+        op_list.append(
+            qml.adjoint(qml.TemporaryAnd([x_wires_pl[0], y_wires_pl[0], work_wires_pl[0]]))
+        )
+
+        op_list.append(qml.CNOT([x_wires_pl[0], y_wires_pl[0]]))
+
+        return op_list
+
+
+def _semiadder_resources(num_y_wires):
+    # In the case where len(x_wires) < len(y_wires), this is an upper bound.
+    return {
+        qml.TemporaryAnd: num_y_wires - 1,
+        qml.decomposition.adjoint_resource_rep(qml.TemporaryAnd, {}): num_y_wires - 1,
+        qml.CNOT: 6 * (num_y_wires - 2) + 3,
```

Comment:
Do we take into account `_right_operator` and `_left_operator` here, or do we not need to?

Reply:
It is include 👍